Optimize `IN` performance with specialized implementations

geoffreyclaude opened the pull request:
## Status

This aggregate PR has been superseded by the split stacked series for #19241.

The review path is now:

1. #21927: generic static-filter cleanup and fallback optimization
2. #23011: Bitmap filter for UInt8
3. #23012: Bitmap filter for UInt16
4. #23013: zero-copy reinterpretation for 1-byte and 2-byte bitmap paths
5. #23014: branchless filter for small primitive lists
6. #23015: direct-probe hash filter for large primitive lists
7. #23016: string-view filters for Utf8View and BinaryView
8. #23017: two-stage filter for legacy Utf8 and LargeUtf8
9. #23018: zero-copy reinterpretation for FixedSizeBinary

Each PR now owns one focused step in the optimization stack, with its own explanation and CI signal. Closing this aggregate PR avoids duplicate review on the same work.

#19241 remains the umbrella issue for the overall `IN LIST` performance work.

geoffreyclaude commented:
run benchmark in_list

Dandandan commented on datafusion/physical-expr/src/expressions/in_list/array_filter.rs:
```diff
@@ -0,0 +1,159 @@
+    ///
+    /// Note: usize::hash is not used, instead the raw entry
+    /// API is used to store entries w.r.t their value
+    map: HashMap<usize, (), ()>,
```
Probably better to use the `HashTable` API

geoffreyclaude replied:
Updated to either HashTable or HashSet, depending on the underlying type

Dandandan commented on datafusion/physical-expr/src/expressions/in_list/array_filter.rs:
```diff
@@ -0,0 +1,159 @@
+                        false => Some(negated),
+                    }
+                })
+                .collect())
```
It's more efficient to collect nulls / values separately.

Dandandan commented:
run benchmark tpch tpcds

geoffreyclaude commented:
run benchmark in_list

geoffreyclaude commented:
run benchmark in_list

geoffreyclaude commented:
run benchmark in_list

geoffreyclaude commented:
run benchmark in_list

geoffreyclaude committed:
Implement Direct Probe (Hash) Filter for large primitive lists

Implements a fast hash table using open addressing with linear probing and a 25% load factor. Replaces the legacy HashSet for primitives, reducing indirection. Triggers for primitives when list size exceeds branchless thresholds.

geoffreyclaude committed:
Implement String View (Utf8View/BinaryView) Optimizations

Introduces a two-stage filter for ByteView types. Stage 1 uses a fast DirectProbeFilter on masked views (len + prefix) for quick rejection; Stage 2 performs full verification only for potential long-string matches. Triggers for Utf8View and BinaryView.

geoffreyclaude committed:
Extend Bitmap Filter to UInt16 (Heap-based)

Implements an 8 KB heap-allocated bitmap for UInt16. Maintains O(1) performance while handling the larger value space. Triggers for UInt16 arrays.

geoffreyclaude committed:
Implement Branchless Filter for small primitive lists

Adds a const-generic unrolled comparison chain that avoids CPU branching. Outperforms hash lookups for very small lists. Triggers for primitives when list size <= 32 (4-byte), 16 (8-byte), or 4 (16-byte).

geoffreyclaude committed:
Extend Bitmap Filter to UInt16 (Heap-based)

Implements an 8 KB heap-allocated bitmap for UInt16. Maintains O(1) performance while handling the larger value space. Triggers for UInt16 arrays.

geoffreyclaude committed:
Implement Zero-Copy Reinterpretation and enable Int8/Int16 Bitmaps

Introduces zero-copy buffer reinterpretation to allow signed integers and other 1 or 2-byte primitive types (e.g. Float16) to use the high-performance bitmap filters. Triggers for all types with 1-byte or 2-byte width.

geoffreyclaude committed:
Implement Legacy String Optimization (Utf8TwoStageFilter)

Port of the two-stage View optimization to standard Utf8 and LargeUtf8 types. Encodes strings as i128 (len + prefix) for fast O(1) pre-filtering before falling back to full string comparison. Triggers for Utf8 and LargeUtf8.

geoffreyclaude committed:
Implement Zero-Copy Reinterpretation and enable Int8/Int16 Bitmaps

Introduces zero-copy buffer reinterpretation to allow signed integers and other 1 or 2-byte primitive types (e.g. Float16) to use the high-performance bitmap filters. Triggers for all types with 1-byte or 2-byte width.

geoffreyclaude committed:
Implement Legacy String Optimization (Utf8TwoStageFilter)

Port of the two-stage View optimization to standard Utf8 and LargeUtf8 types. Encodes strings as i128 (len + prefix) for fast O(1) pre-filtering before falling back to full string comparison. Triggers for Utf8 and LargeUtf8.

geoffreyclaude committed:
Implement Bitmap Filter for UInt8 (Stack-based)

Replaces HashSet<u8> with a 32-byte stack-allocated bitmap. Provides O(1) membership testing via bit-shifting, significantly reducing memory overhead and improving cache locality. Triggers for UInt8 arrays.

geoffreyclaude committed:
Implement String View (Utf8View/BinaryView) Optimizations

Introduces a two-stage filter for ByteView types. Stage 1 uses a fast DirectProbeFilter on masked views (len + prefix) for quick rejection; Stage 2 performs full verification only for potential long-string matches. Triggers for Utf8View and BinaryView.

geoffreyclaude committed:
Implement Direct Probe (Hash) Filter for large primitive lists

Implements a fast hash table using open addressing with linear probing and a 25% load factor. Replaces the legacy HashSet for primitives, reducing indirection. Triggers for primitives when list size exceeds branchless thresholds.

geoffreyclaude committed:
Implement Bitmap Filter for UInt8 (Stack-based)

Replaces HashSet<u8> with a 32-byte stack-allocated bitmap. Provides O(1) membership testing via bit-shifting, significantly reducing memory overhead and improving cache locality. Triggers for UInt8 arrays.

geoffreyclaude committed:
Implement Branchless Filter for small primitive lists

Adds a const-generic unrolled comparison chain that avoids CPU branching. Outperforms hash lookups for very small lists. Triggers for primitives when list size <= 32 (4-byte), 16 (8-byte), or 4 (16-byte).

geoffreyclaude commented:
run benchmark in_list

geoffreyclaude commented:
run benchmark in_list

geoffreyclaude commented:
run benchmark in_list

geoffreyclaude committed:
Implement FixedSizeBinary zero-copy reinterpretation optimization

FixedSizeBinary(N) arrays share the same contiguous buffer layout as primitive arrays, so for power-of-2 widths (1, 2, 4, 8, 16) we can zero-copy reinterpret them and use the optimized primitive filters (bitmap, branchless, hash) instead of falling through to the NestedTypeFilter fallback.

geoffreyclaude committed:
Implement FixedSizeBinary zero-copy reinterpretation optimization

FixedSizeBinary(N) arrays share the same contiguous buffer layout as primitive arrays, so for power-of-2 widths (1, 2, 4, 8, 16) we can zero-copy reinterpret them and use the optimized primitive filters (bitmap, branchless, hash) instead of falling through to the NestedTypeFilter fallback.

adriangb commented:
@alamb @geoffreyclaude I wonder if you have any thoughts on how we can move forward with this.

May a high level the tradeoff I see is:
- This adds complexity. It’s localized and self contained but still.
- This meaningfully improves performance of an important operator

Practically speaking I want to get this across the line but I find this hard to review, both because of the size of the diff but also because of the nuance of the implementation. Quite honestly without AI explaining to me some of the more niche branches I’d struggle to grok this or find holes in the logic.

geoffreyclaude commented:
@adriangb Thanks for looking into this a bit more!

If we agree most of these optimizations are worth pursuing (my biased Datadog opinion being that the Int8/Int16 and Utf8 optims are the most critical), we should first get the first two commits merged. Those set the foundations for the rest. Each subsequent optimization commit is then much easier to reason about independently and they can be easily split into dedicated PRs.

The first commit is "just" AI generated benchmarks, I don't think we need to care too much about that one. The second one though is the most important...

alamb commented:
Adding new benchmarks in a separate PR sounds like a good way to start

Also a related PR here:
- https://github.com/apache/datafusion/pull/21547

adriangb commented:
Yeah agreed. @geoffreyclaude if you make a PR with just the benchmarks we can merge that immediately. Then the second commit would be its own PR. The rest I feel like we really need to look at the tradeoff of performance vs. complexity one by one and e.g. bake them off against https://github.com/apache/datafusion/pull/21547.

geoffreyclaude commented:
> Yeah agreed. @geoffreyclaude if you make a PR with just the benchmarks we can merge that immediately. Then the second commit would be its own PR. The rest I feel like we really need to look at the tradeoff of performance vs. complexity one by one and e.g. bake them off against https://github.com/apache/datafusion/pull/21547.

@adriangb I (well, Codex...) cherry-picked the first two commits to dedicated PRs:
- https://github.com/apache/datafusion/pull/21648
- https://github.com/apache/datafusion/pull/21649

geoffreyclaude committed:
Optimize generic InList static filtering

Introduces NestedTypeFilter for non-primitive constant IN lists. Replaces the legacy ArrayStaticFilter fallback with a HashTable-backed lookup and shared bitmap result construction.

geoffreyclaude commented:
run benchmark in_list_strategy

geoffreyclaude committed:
Refactor generic InList static filter helpers

geoffreyclaude committed:
Build InList results from bitmaps

geoffreyclaude committed:
Optimize generic InList static filtering

adriangb commented:
@geoffreyclaude I spoke with @alamb today, we think what we should do here is wait until DataFusion 54 is released and then merge this given all tests pass and benchmark numbers look good (which they do).

geoffreyclaude commented:
@adriangb Sorry for the late reply, I was off and somehow missed this notification. I agree that this change is pretty well tested and benchmarked now, so I'd be pretty surprised if it introduces major issues (we might still have some weird edge case of course, as usual.)
What I'm a bit more concerned about is maintainability as it does introduce quite a bit of complex features. In particular, "Commit 7: Implement Direct Probe", while showing significant gains, is pretty tricky to understand well. I would never have come up with something like this without AI "helping" out.
WDYT? Are the performance gains worth the additional future headaches when new work in this area is needed?

geoffreyclaude commented:
run benchmark in_list_strategy

adriangb commented:
> @adriangb Sorry for the late reply, I was off and somehow missed this notification. I agree that this change is pretty well tested and benchmarked now, so I'd be pretty surprised if it introduces major issues (we might still have some weird edge case of course, as usual.) What I'm a bit more concerned about is maintainability as it does introduce quite a bit of complex features. In particular, "Commit 7: Implement Direct Probe", while showing significant gains, is pretty tricky to understand well. I would never have come up with something like this without AI "helping" out. WDYT? Are the performance gains worth the additional future headaches when new work in this area is needed?

That is my concern as well :)

It's hard for me to judge what is complex because I've never seen it before / don't have a CS degree vs. is complex for anyone. I'm prepared to basically trust your judgment on it.

How about this test for each commit / optimization: if your team came and said "hey we are hitting an `InList` bug in prod, please go fix it" and you had never seen the code before, would you feel "this is complicated but the X% perf gain makes it worth it" or "my first instinct is to rip this out it's not worth it". Maybe we pick 1-2 lowest "bang for the buck" optimizations in here and stash those for later?

One thing we could do is merge as stacked PRs so that commits don't get squashed and it's easier to revert individual pieces. Ultimately if there were issues because this is self contained I think the initial answer would be to revert the problematic optimization.

geoffreyclaude commented:
> > @adriangb Sorry for the late reply, I was off and somehow missed this notification. I agree that this change is pretty well tested and benchmarked now, so I'd be pretty surprised if it introduces major issues (we might still have some weird edge case of course, as usual.) What I'm a bit more concerned about is maintainability as it does introduce quite a bit of complex features. In particular, "Commit 7: Implement Direct Probe", while showing significant gains, is pretty tricky to understand well. I would never have come up with something like this without AI "helping" out. WDYT? Are the performance gains worth the additional future headaches when new work in this area is needed?
> 
> That is my concern as well :)
> 
> It's hard for me to judge what is complex because I've never seen it before / don't have a CS degree vs. is complex for anyone. I'm prepared to basically trust your judgment on it.
> 
> How about this test for each commit / optimization: if your team came and said "hey we are hitting an `InList` bug in prod, please go fix it" and you had never seen the code before, would you feel "this is complicated but the X% perf gain makes it worth it" or "my first instinct is to rip this out it's not worth it". Maybe we pick 1-2 lowest "bang for the buck" optimizations in here and stash those for later?
> 
> One thing we could do is merge as stacked PRs so that commits don't get squashed and it's easier to revert individual pieces. Ultimately if there were issues because this is self contained I think the initial answer would be to revert the problematic optimization.

@adriangb Good idea, stacked PRs is probably the safest bet here. We don't want all commits to get squashed into a single one (is that the merge policy on `datafusion`?), impossible to revert out of selectively. With stacked PRs, each optimization gets its own dedicated PR/commit, easy to revert and git bisect in case of issue (manually or with coding agents.)
The most complex commit I mentioned above does have some pretty significant gains, so it would be a shame to not take advantage of it! I'll work on stacking these properly (but maybe we can skip the in-depth review of each and focus on the big picture here.)

adriangb commented:
> is that the merge policy on datafusion?

I believe so. We only have 1 merge button and afaik it squashes.

> maybe we can skip the in-depth review of each and focus on the big picture here

Yep agreed. I think if benchmarks look good, all tests pass and there are no public API changes we can move forward with them.

alamb commented:
Thsoe are some pretty sweet performance results. I will try and find some time to review this more carefully

geoffreyclaude commented:
@adriangb @alamb In case you missed it, I've closed this PR and rewritten it as a stack of multiple PRs (one per optim.) The root issue https://github.com/apache/datafusion/issues/19241 links them all in order. Due to github limitations, I can't do proper stacked PRs so each PR has all the previous commits in it. But I'll rebase over main as we merge them in order, which will clean them up.

alamb commented:
Thank you very much for doing this in steps -- I will help move them along

I was finding it super challenging to find enough contiguous focus time for this PR. I am feeling very good about the stacked PR approahc -- I'll try and help
